feat(metrics-extraction): Add widget columns to extraction

k-fish opened the pull request:
### Summary
This adds dashboard widget columns to the MetricSpec's emitted from the relay metric extraction code. Moving forward this should mean any widget with columns should start working with on-demand metrics (once the endpoint code has been written).

Also in this PR is cardinality checking functionality to block emitting any metrics specs that exceeds a high upper bound. We will also be limiting cardinality to a lower number upon dashboard creation, this is mostly to minimize churn on existing dashboards.

iambriccardo reviewed (commented):
Thanks for the contribution. Will you follow up with another PR for the querying? IDK, if there is a need to do some modifications in the query builders since I lack knowledge on how they handle multi-aggregate queries.

iambriccardo commented on src/sentry/snuba/metrics/extraction.py:
```diff
@@ -905,12 +936,28 @@ def tags_conditions(self, project: Project) -> List[TagSpec]:
+        field = _map_field_name(column)
+
+        return {
+            "key": field,
```
Another small thing, if we keep the column name as the key, we might leverage the column resolution of the query builders but I am not 100% sure it could be plugged in easily. 

How you did it it's perfectly fine I am just thinking out loud.

k-fish replied:
So I was thinking keeping key=field so we know exactly was attempted for extraction into the field _but_ I think you're right since they we don't update the QueryBuilders to just use the intended column. Also we'll be making a new set of indexer values when a lot of them already exist for MEP data.

iambriccardo replied:
That's a good idea. At the end of the day we will have our own custom logic for the query resolution, so it's not a big deal.

Actually, it's even better since in a nicer future we would like to lift out on-demand metrics out of the query builder but this is a utopian future.

k-fish commented on src/sentry/relay/config/metric_extraction.py:
```diff
@@ -265,13 +281,77 @@ def _convert_widget_query_to_metric(
+
+    max_cardinality_allowed = options.get("on_demand.max_widget_cardinality.count")
+    cache_key = f"check-widget-query-cardinality:{widget_query.id}"
+    cardinality_allowed = cache.get(cache_key)
```
Memcache on prod

k-fish commented on src/sentry/relay/config/metric_extraction.py:
```diff
@@ -265,13 +281,77 @@ def _convert_widget_query_to_metric(
+        sentry_sdk.capture_exception(error)
+        return False
+
+    cache.set(cache_key, True)
```
It'll be the default here, which is low. Thinking about this, not only do we need to set it high, we need to add some jitter to avoid thundering herd once we have more and more of these since they are making discover queries.

k-fish replied:
Set it to 1d + 25% jitter. We probably want it even higher in practice for GA, but this should suffice for LA.

armenzg replied:
I see timeout being set on the other cache.set() calls; had you intended to add it here as well?

k-fish commented on src/sentry/relay/config/metric_extraction.py:
```diff
@@ -265,13 +281,77 @@ def _convert_widget_query_to_metric(
+        "project_objects": [project],
+        "organization_id": project.organization_id,  # Organization id has to be specified to not violate allocation policy.
+    }
+    params["statsPeriod"] = "1d"
+    start, end = get_date_range_from_params(params)
+    params["start"] = start
+    params["end"] = end
```
Hmm I suppose it doesn't matter going with a hardcoded time in this case since the statsPeriod is internal.

k-fish commented on src/sentry/snuba/metrics/extraction.py:
```diff
@@ -310,8 +310,10 @@ def should_use_on_demand_metrics(
     aggregate: str,
     query: Optional[str],
     prefilling: bool = False,
+    columns: Optional[Sequence[str]] = None,
```
sure :+1:

k-fish commented on src/sentry/snuba/metrics/extraction.py:
```diff
@@ -394,15 +399,18 @@ def _get_percentile_op(args: Sequence[str]) -> Optional[MetricOperationType]:
         return SupportedBy(on_demand_metrics=True, standard_metrics=False)
 
-    arg = args[0]
+    arg = columns[0]
```
Refactored, what it's really doing is the same as checking whether the arg in a function is compatible, which basically checks whether a 'column' is compatible since functions take a column as a arg in some cases.

I did it for parity here but looking at this again I think `columns` should all be checked for support, and the result joined. Not sure why we're getting the first index only here.

k-fish replied:
For clarity I think I'm re-arranging this so the aggregate args and groupbys are checked separately.

k-fish commented on src/sentry/snuba/metrics/extraction.py:
```diff
@@ -891,6 +918,10 @@ def _query_for_hash(self):
         return str(_deep_sorted(self.condition))
 
+    def _columns_for_hash(self):
+        # A sorted list of columns (for idempotency) for the hash, since groupbys will be unique per on_demand metric.
```
right right

k-fish commented:
@iambriccardo yeah the querybuilder changes need to be done for both [tables](https://github.com/orgs/getsentry/projects/156/views/2?pane=issue&itemId=43135827) and [timeseries](https://github.com/orgs/getsentry/projects/156/views/2?pane=issue&itemId=43136370) 

Want to take another look?
- I switched all mentions of 'column' to 'groupby' since it's pretty confusing what a column is, groupby is a better descriptor. 
- Added a cache TTL

iambriccardo commented:
> @iambriccardo yeah the querybuilder changes need to be done for both [tables](https://github.com/orgs/getsentry/projects/156/views/2?pane=issue&itemId=43135827) and [timeseries](https://github.com/orgs/getsentry/projects/156/views/2?pane=issue&itemId=43136370)
> 
> Want to take another look?
> 
> * I switched all mentions of 'column' to 'groupby' since it's pretty confusing what a column is, groupby is a better descriptor.
> * Added a cache TTL

Thanks so much for renaming, I think that yeah column was a bit confusing as a technical term, since as you say, we are running an aggregate over a group by composed of columns. I actually had to check how columns were stored in the db to figure out how they were working.

iambriccardo reviewed (approved):
LGTM, thanks again for the help!

jjbayer commented on src/sentry/relay/config/metric_extraction.py:
```diff
@@ -265,13 +285,80 @@ def _convert_widget_query_to_metric(
+    if cardinality_allowed is not None:
+        return cardinality_allowed
+
+    unique_columns = [f"count_unique({column})" for column in widget_query.columns]
+
+    query_builder = QueryBuilder(
+        dataset=Dataset.Discover,
+        params=params,
+        selected_columns=unique_columns,
+        config=QueryBuilderConfig(
+            transform_alias_to_input_format=True,
+        ),
+    )
```
Do we know roughly how expensive this query is, and how often it will run? Any risk of hitting a timeout here?

k-fish replied:
Yes, I'm going to be wrapping this with another exception so the config still generates, on top of that I'm going to stagger the queries by mod'ing the widget query id.
